client/proxy.cc

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2016 SUSE LLC
+ *
+ * All Rights Reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of version 2 of the GNU General Public License as published
+ * by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact Novell, Inc.
+ *
+ * To contact Novell about this file by physical or electronic mail, you may
+ * find current contact information at www.novell.com.
+ */
+
+
+#include <sstream>
+#include <iostream>
+#include <algorithm>
+
+#include <snapper/AppUtil.h>
+#include <snapper/SnapperDefines.h>
+
+#include "utils/text.h"
+#include "proxy.h"
+
+
+using namespace std;
+
+
+string
+ProxyConfig::getSubvolume() const
+{
+    string subvolume;
+    getValue(KEY_SUBVOLUME, subvolume);
+    return subvolume;
+}
+
+
+bool
+ProxyConfig::getValue(const string& key, string& value) const
+{
+    map<string, string>::const_iterator it = values.find(key);
+    if (it == values.end())
+	return false;
+
+    value = it->second;
+    return true;
+}
+
+
+ProxySnapshots::iterator
+ProxySnapshots::find(unsigned int num)
+{
+    return find_if(proxy_snapshots.begin(), proxy_snapshots.end(),
+		   [num](const ProxySnapshot& x) { return x.getNum() == num; });
+}
+
+
+ProxySnapshots::const_iterator
+ProxySnapshots::find(unsigned int num) const
+{
+    return find_if(proxy_snapshots.begin(), proxy_snapshots.end(),
+		   [num](const ProxySnapshot& x) { return x.getNum() == num; });
+}
+
+
+ProxySnapshots::iterator
+ProxySnapshots::findNum(const string& str)
+{
+    istringstream s(str);
+    unsigned int num = 0;
+    s >> num;
+
+    if (s.fail() || !s.eof())
+    {
+	cerr << sformat(_("Invalid snapshot '%s'."), str.c_str()) << endl;
+	exit(EXIT_FAILURE);
+    }
+
+    iterator ret = find(num);
+    if (ret == proxy_snapshots.end())
+    {
+	cerr << sformat(_("Snapshot '%u' not found."), num) << endl;
+	exit(EXIT_FAILURE);
```

Comment:
How about returning an end() iterator here?

While the above exit() can be explained with an I/O error that is hard to recover from, this might be a simple user error (typo, for example). If one day there is something like a snapper shell (you'll never know), this error could be presented to the user so he can fix it and try with another number. Calling exit() immeditaly here would make this impossible.